sql/plan/sliding_range.go

```diff
@@ -0,0 +1,183 @@
+	rows         []sql.Row
+	err          error
+}
+
+var _ sql.Node = (*SlidingRange)(nil)
+
+func NewSlidingRange(child sql.Node, lhsSchema sql.Schema, rhsSchema sql.Schema, value, min, max string, rangeIsClosedBelow, rangeIsClosedAbove bool) (*SlidingRange, error) {
+	// TODO: IndexOfColName is Only safe for schemas corresponding to a single table, where the source of the column is irrelevant.
+	maxColumnIndex := rhsSchema.IndexOfColName(max)
+	newSr := &SlidingRange{
+		activeRanges:       priorityQueue{},
+		pendingRow:         nil,
+		valueColumnIndex:   lhsSchema.IndexOfColName(value),
+		minColumnIndex:     rhsSchema.IndexOfColName(min),
+		maxColumnIndex:     maxColumnIndex,
+		comparisonType:     rhsSchema[maxColumnIndex].Type,
+		rangeIsClosedBelow: rangeIsClosedBelow,
+		rangeIsClosedAbove: rangeIsClosedAbove,
+	}
+	newSr.Child = child
+	newSr.activeRanges.slidingRange = newSr
+	return newSr, nil
+}
+
+func (s *SlidingRange) String() string {
+	return s.Child.String()
+}
+
+func (s *SlidingRange) WithChildren(children ...sql.Node) (sql.Node, error) {
+	if len(children) != 1 {
+		return nil, fmt.Errorf("ds")
+	}
+
+	s2 := *s
+	s2.UnaryNode = UnaryNode{Child: children[0]}
+	return &s2, nil
+}
+
+func (s *SlidingRange) CheckPrivileges(ctx *sql.Context, opChecker sql.PrivilegedOperationChecker) bool {
+	return s.Child.CheckPrivileges(ctx, opChecker)
+}
+
+var _ sql.Node = (*SlidingRange)(nil)
+
+func (s *SlidingRange) Initialize(ctx *sql.Context, childRowIter sql.RowIter) (err error) {
+	s.childRowIter = childRowIter
+	s.activeRanges = priorityQueue{
+		slidingRange: s,
+		rows:         nil,
+		err:          nil,
+	}
+	s.pendingRow, err = childRowIter.Next(ctx)
+	return err
+}
+
+func (s *SlidingRange) IsInitialized() bool {
+	return s.childRowIter != nil
+}
+
+func (s *SlidingRange) AcceptRow(ctx *sql.Context, row sql.Row) (sql.RowIter, error) {
+	// Remove rows from the heap if we've advanced beyond their max value.
+	for s.activeRanges.Len() > 0 {
+		maxValue := s.activeRanges.Peek()
+		compareResult, err := s.comparisonType.Compare(row[s.valueColumnIndex], maxValue)
+		if err != nil {
+			return nil, err
+		}
+		if (s.rangeIsClosedAbove && compareResult > 0) || (!s.rangeIsClosedAbove && compareResult >= 0) {
+			heap.Pop(&s.activeRanges)
+		} else {
+			break
+		}
+	}
+
+	// Advance the child iterator until we encounter a row whose min value is beyond the range.
+	for s.pendingRow != nil {
+		minValue := s.pendingRow[s.minColumnIndex]
+		compareResult, err := s.comparisonType.Compare(row[s.valueColumnIndex], minValue)
+		if err != nil {
+			return nil, err
+		}
+
+		if (s.rangeIsClosedBelow && compareResult < 0) && (!s.rangeIsClosedBelow && compareResult <= 0) {
+			break
+		} else {
+			heap.Push(&s.activeRanges, s.pendingRow)
+		}
+
+		s.pendingRow, err = s.childRowIter.Next(ctx)
+		if err != nil {
+			if errors.Is(err, io.EOF) {
+				// We've already imported every range into the priority queue.
+				s.pendingRow = nil
```

Comment:
i'd be surprised if this line was necessary

Reply:
It's needed for the case where the RHS table is exhausted before the LHS table. We still need to accept every remaining LHS row, and maintain the heap, but we need to track the fact that there's no RHS rows to process.